### ISSUE TITLE
directory tree portal

### ISSUE THREAD

**Issue by mikelpr** [flatpak/xdg-desktop-portal#200]
file chooser portal for selecting a directory and its contents + subdirectories for RW

**Comment by matthiasclasen** [comment]
This is not something that the document portal lets us do (export a tree).

**Comment by mjog** [comment]
Maybe a dupe of #158?

**Comment by callaa** [comment]
I also have a use case that would benefit from a directory portal. My application can export animations, either as video files or as individual frames. When exporting frames, it's not practical to prompt the user for each individual frame, so being able to open a directory is essential. In this particular use case, subdirectory access is not necessary.

Applications like Darktable would also benefit from being able to open directories, both for browsing images and for batch exporting processed photos.

**Comment by Exalm** [comment]
If we're gathering use cases, GNOME Games needs this to be able to support DOS games without full file system access. Those games expect to be able to work with surrounding files, e.g. for reading/writing config and saves.

**Comment by TingPing** [comment]
It is known hundreds of applications can use this. Lets keep this issue for details on how to implement it.

**Comment by swick** [comment]
What's the situation here. Is there a clear path forward?

Showing a GUI to add a filesystem override would be easy but requires an app restart to become visible which is not nice at all.
Bind-mounting something into a running container doesn't really seem possible with the tools the kernel gives us right now.
Yet another fuse system might work.

More ideas?

**Comment by mjog** [comment]
Portal API support for this landed in xdg-desktop-portal 1.5.3 and xdg-desktop-portal-gtk 1.5.1. Not sure about the KDE portal.

Applications can use the new API directly or use libportal to access it (assuming people have the portal versions above installed), but no transparent support has landed in GTK or Qt as of yet.

**Comment by swick** [comment]
Do you mean the SaveFiles method on the document portal? It's the only thing I could find which remotely has anything to do with this issue but it certainly isn't a complete solution to it.

**Comment by mjog** [comment]
> Do you mean the SaveFiles method on the document portal?

Oh, yeah I see how that's probably not helpful here. Apologies for the noise.

**Comment by swick** [comment]
Alright, I remembered reading something about a new kernel interface for mounting.

https://lwn.net/Articles/759499/
https://lwn.net/Articles/809125/

This might be suitable for adding new bind mounts in a running container. First I thought it's possible to create a bind mount using `fsopen` and `fsconfig` like so

```
sfd = fsopen(fstype, 0);
fsconfig(sfd, FSCONFIG_SET_FLAG, "ms_bind", NULL, 0);
fsconfig(sfd, FSCONFIG_SET_STRING, "source", "/host/path", 0);
fsconfig(sfd, FSCONFIG_CMD_CREATE, NULL, NULL, 0);
mfd = fsmount(sfd, 0, 0);
```
but apparently that doesn't work ("Block device required"; maybe it's possible and I'm just doing something wrong).
Another I had was to use the old `mount` syscall to create a bind mount on the host and then use `open_tree`.
Either way should yield a file descriptor that represents the mount and can eventually be inserted into the fs tree by calling `move_mount`. I believe this should work across mount namespaces but I'm really not sure because I'm failing at doing all of this in the host namespace already.

I'd really appreciate if someone who knows more about this would chime in.

**Comment by swick** [comment]
Had some more time to test and…

host:
```
mfd = open_tree(AT_FDCWD, host_directory, OPEN_TREE_CLONE);
send_to_sandbox(mfd);
```
sandbox:
```
mfd = recv_from_host();
move_mount(mfd, "", AT_FDCWD, sandbox_directory, MOVE_MOUNT_F_EMPTY_PATH);
```
This actually works and creates a bind mount from the host's host_directory to the running container's sandbox_directory. It however requires mount caps on both the host and the container.

**Comment by swick** [comment]
The mount caps are making the previous approach basically useless but I've found a new way using slave mounts which is looking promising (https://github.com/containers/bubblewrap/pull/356).

**Comment by Erick555** [comment]
@swick are you aware of recent work? https://github.com/alexlarsson/xdg-desktop-portal/commits/new-fuse-backend

**Comment by swick** [comment]
No, I wasn't. That's unfortunate timing.

**Comment by hadess** [comment]
This [GTK change](https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/1523) (and related changes in xdg-desktop-portal and xdg-desktop-portal-gtk) should allow selecting directories for things like music libraries, or output directory if an application is expected to output everything to a single directory.

The bug was fixed in c616b491fe0ac67f78a14bcaf0d3a150bfadfd82 (and further fixed still in some opened MRs).

We're still however missing a UI that would make it clear that opening a file, eg. a project file, would also give access to the rest of the files in that same directory. I've moved the conversation to https://github.com/flatpak/xdg-desktop-portal/issues/463
